### PR TITLE
Add FieldIndex implementation

### PR DESCRIPTION
**Pull request by wu-hui**

### CODE REVIEW

**Review by ehsannas — APPROVED** [review]
Generally LGTM. Some nits inlined.

File: Firestore/core/src/model/document.h [hunk 1]
```diff
@@ -52,6 +52,10 @@ class Document {
     return document_.ToString();
   }
 
+  SnapshotVersion read_time() const {
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply Firestore/core/src/model/document.h]: Done.

File: Firestore/core/src/model/field_index.cc [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+util::ComparisonResult IndexOffset::CompareTo(const IndexOffset& rhs) const {
+  auto result = read_time_.CompareTo(rhs.read_time());
+  if (result != util::ComparisonResult::Same) {
+    return result;
+  }
+
+  return document_key_.CompareTo(rhs.document_key());
+}
+
+util::ComparisonResult IndexOffset::DocumentCompare(const Document& a,
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply Firestore/core/src/model/field_index.cc]: It is in the `IndexOffset` inner class of `FieldIndex` for Android.

File: Firestore/core/src/model/field_index.h [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+ public:
+  /**
+   * Creates an offset that matches all documents with a read time higher than
+   * `read_time` or with a key higher than `key` for equal read times.
+   */
+  IndexOffset(SnapshotVersion read_time, DocumentKey key)
+      : read_time_(std::move(read_time)), document_key_(std::move(key)) {
+  }
+
+  static IndexOffset None();
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply Firestore/core/src/model/field_index.h:87]: Done.

File: Firestore/core/src/model/field_index.h [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+ * ArrayContains/ArrayContainsAny queries.
+ *
+ * Unlike the backend, the SDK does not differentiate between collection or
+ * collection group-scoped indices. Every index can be used for both single
+ * collection and collection group queries.
+ */
+class FieldIndex {
+ public:
+  /** An ID for an index that has not yet been added to persistence. */
+  static int32_t UnknownId() {
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply Firestore/core/src/model/field_index.h]: I like constexpr better here, thanks.

File: Firestore/core/src/model/field_index.h [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+    return -1;
+  }
+
+  /**
+   * The initial sequence number for each index. Gets updated during index
+   * backfill.
+   */
+  static ListenSequenceNumber InitialSequenceNumber() {
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply Firestore/core/src/model/field_index.h]: Done.

File: Firestore/core/src/model/field_index.h [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+  }
+
+  /** Returns all directional (ascending/descending) segments for this index. */
+  std::vector<Segment> GetDirectionalSegments() const;
+
+  /** Returns the ArrayContains/ArrayContainsAny segment for this index. */
+  absl::optional<Segment> GetArraySegment() const;
+
+ private:
+  int32_t index_id_ = UnknownId();
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply Firestore/core/src/model/field_index.h:238]: It'd be 0 for a default instance, -1 feels more correct?

File: Firestore/core/test/unit/model/field_index_test.cc [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+  EXPECT_EQ(ComparisonResult::Ascending,
+            FieldIndex::SemanticCompare(original, different));
+}
+
+TEST(FieldIndexTest, ComparatorIgnoresIndexId) {
+  FieldIndex original = MakeFieldIndex("collA", 1, FieldIndex::InitialState());
+  FieldIndex same = MakeFieldIndex("collA", 1, FieldIndex::InitialState());
+  FieldIndex different = MakeFieldIndex("collA", 2, FieldIndex::InitialState());
+  EXPECT_EQ(ComparisonResult::Same,
+            FieldIndex::SemanticCompare(original, same));
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply Firestore/core/test/unit/model/field_index_test.cc]: Done.

**Review by ehsannas — APPROVED** [review]

**Pull request merged**